### ISSUE TITLE
Settings for controlling if outputs should be saved to disk 

### ISSUE THREAD

**Issue by DonJayamanne** [microsoft/vscode#124551]
Originally filed here by @rebornix https://github.com/microsoft/vscode-jupyter/issues/4670

I think we should consider adding this feature into VSCode. I've implemented this same feature in Kusto notebooks.
Feels like something generic for VS Code.

Similar to having a cell toolbar position per viewtype, we can have `Save Notebook Outputs` per viewtype. This can be achieved easily from VS Code by passing an empty array into the serializer, when saving notebooks.

I personally think this would be useful for rest notebooks as well, one might want the output from the REST & GithubIssues Notebook to be saved in the notebook, just like kusto notebooks (some might want the charts/data & some might not). 

Thus all notebooks can be considered serializable (including output **for sharing**) & user choses whether the output is to be persisted on disc or not.

E.g. with .NET team building their notebook format (*.dib) this would have to be impelemented yet again in their extension.

@rebornix /cc

**Comment by rebornix** [comment]
> E.g. with .NET team building their notebook format (*.dib) this would have to be impelemented yet again in their extension.

I doubt so, `dib` is a concatenation of cell inputs so it doesn't save outputs.

---

> if outputs should be saved to disk

if we are talking about serializing the content and saving to disk, then it's the responsibility of the content serializer. VS Code has no idea how the outputs will be used to generate the final file format. If we ever implement something in the core, there are many open questions:

* should the content serializer completely drop `outputs` info in the file, or save them as `outputs: []`? how would it know it's empty because users disable it or it's simply empty?
* should the content serializer save execution count? Same question as above, how would it know?
* is there any other metadata that can be linked with `outputs` and should have the same persistence story as `outputs`?

In Kusto extension, IIRC if users set not to save outputs, then we also make the `executionCount` transient. Maybe we do the same in Jupyter, but we don't know yet if it's generic enough.

cc @jrieken @tanhakabir @roblourens

**Comment by jrieken** [comment]
Yeah, I don't believe that it can be done. The closest request we have is support for transient per output-instance, not per notebook type: https://github.com/microsoft/vscode/issues/120600

**Comment by tanhakabir** [comment]
I don't have this case where I'm not saving outputs yet (no request yet) but this case might be possible from core by not giving output cell items in NotebookData when the serialize function is called?

So far though I don't have such a feature planned

**Comment by andreaschiappacasse** [comment]
+1 
This would be also super useful in managing version control of .ipynb, in order to avoid committing large outputs/sesntive output which may be generated by the notebook.

**Comment by MrNickArcher** [comment]
I came looking for this. Sad to see it hasn't got much enthusiasm. I also want it

- Preparing notebooks for teaching without accidentally giving students any pre-run outputs
- Hiding sensitive output and metadata when committing to git repositories

**Comment by RatanRSur** [comment]
Chiming in to say I'd also like this feature

**Comment by PiotrCzapla** [comment]
It would be best to expose an ability to filter what get's saved, [NBDEV](https://nbdev.fast.ai) that facilitate building python libraries in notebooks in literate programming, has special comment markup to remove only some outputs.  I would love to be able to run the same hook that is already implemented for Jupyter in vscode.

**Comment by JotaRata** [comment]
Does anybody know if this can be made possible by using the new **Notebook: Code Actions On Save** setting?

**Comment by NickCrews** [comment]
Workaround: add a git filter so that in the middle of committing, the notebook is intercepted and the outputs are removed before they are saved by git. The outputs are still saved to disk locally, so you can still see them, but when you git push the outputs won't be included, or if you git checkout a different commit, you won't see the outputs there.

[See this gist](https://gist.github.com/33eyes/431e3d432f73371509d176d0dfb95b6e) (PS, run `git add --renormalize .` as a 4th step before your first commit, so that all of your existing notebooks get scrubbed, or you will get heinous confusing merge conflicts later.)

**Comment by ndgayan** [comment]
This is a big issue when you working on remote SSH with Jupyter Notebook files. During the VSCode's "save" action, a notebook that contains variables/data frames with a large amount of data in the output is written to the remote file. Sometimes this can cause VSCode to crash or wait a few minutes to finish the save action. It would be helpful If you could add an option to disable saving the notebook output to the file and keep it only in the memory during the session.

**Comment by ruiouyangVA** [comment]
I can only find options to clear outputs on staging with git, not upon save. Does anyone know of the aforementioned solution for clearing on save? Thanks!

**Comment by kvablack** [comment]
I would really like this feature as well, for the reason that @ndgayan described. The problem seems to be exacerbated when I'm working on a weak connection (like a mobile hotspot) where the download speed is much faster than the upload speed. I frequently render large images and videos in Jupyter notebooks, making them pretty much unusable unless I clear cell outputs before saving. If I forget even a single time, VSCode locks up on the "saving file" operation and I have to force-reload the window, causing me to lose any changes I made to the cell contents.

**Comment by movy** [comment]
Is this addressed yet? Working on heavy notebooks via VSCode remote is PAIN mostly due to connection being jammed by duplex data streams for saving massive outputs and displaying them back to user... Strangely enough, this is not an issue when very same notebook is open from `jupyter lab` ran on the _same_ remote host. But jupyter lab UX/UI is so outdated, that I still prefer to stick with VSCode despite connection problems, carefully managing amount of outputs, but would love to see this resolved...

**Comment by amunger** [comment]
We do have a setting to limit the size of a notebook that can be backed up: `notebook.backup.sizeLimit`. Setting that lower could help but you would also have to disable auto-save for it to have any real effect.

I'll re-open since this could have a pretty big impact for remote performance.

**Comment by amunger** [comment]
actually, @movy - `"ipynb.experimental.serialization": true` might work a lot better for your case. Keep auto-save enabled and it will use a worker thread to serialize and save the file on the extension host without having to pass it back to the renderer.

If that doesn't work, could you create a new issue for your specific case?

**Comment by movy** [comment]
Actually, I just checked -- and this setting IS enabled, and I did notice lately how snappy saving notebooks has become. Before, saving always caused connection hangups, sometime even disconnect, but now it really became unnoticeable.

However, some libs (plotly express, specifically) do really cause render lag, I will try to switch to plotly.io and if it does not help, I'll post my notebook here. 

Huge thanks to all who works on Jupyter extension for VSCode, as it has been a real life changer, esp. combined with Copilot!

**Comment by DonJayamanne** [comment]
@movy  please can you create a separate issue when you get the info
